### ISSUE TITLE
J4de - The `inflationMultiplier` of L1 bridge may not be the latest when depositing

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-ecoprotocol-judging#17]
J4de

high

# The `inflationMultiplier` of L1 bridge may not be the latest when depositing

## Summary

The `inflationMultiplier` of L1 bridge requires the user to actively obtain it in advance, so its value may not be the latest when depositing

## Vulnerability Detail

```solidity
File: bridge/L1ECOBridge.sol
296     function rebase(uint32 _l2Gas) external {
297         inflationMultiplier = IECO(l1Eco).getPastLinearInflation(
298             block.number
299         );
300
301         bytes memory message = abi.encodeWithSelector(
302             IL2ECOBridge.rebase.selector,
303             inflationMultiplier
304         );
305
306         sendCrossDomainMessage(l2TokenBridge, _l2Gas, message);
307     }
```

`inflationMultiplier` will be updated only after calling `rebase`.

```solidity
File: bridge/L1ECOBridge.sol
333         IECO(_l1Token).transferFrom(_from, address(this), _amount);
334         // gons move across the bridge, with inflation multipliers on either side to correctly scale balances
335 >>      _amount = _amount * inflationMultiplier;
```

And the `inflationMultiplier` is used to multiply the amount when depositing. Therefore the attacker can deposit between the L1 Token update and the L1 bridge update to use the old `inflationMultiplier` to arbitrage.

## Impact

Attackers use old `inflationMultiplier` for arbitrage.

## Code Snippet

https://github.com/eco-association/op-eco/blob/39f205fb46eea3df770f0119a890ffdc1ac8f382/contracts/bridge/L1ECOBridge.sol#L335

## Tool used

Manual Review

## Recommendation

It is recommendation to call `rebase` before deposit.

```diff
    function _initiateERC20Deposit(
        address _l1Token,
        address _l2Token,
        address _from,
        address _to,
        uint256 _amount,
        uint32 _l2Gas,
        bytes calldata _data
    ) internal {
+       rebase();
        // When a deposit is initiated on L1, the L1 Bridge transfers the funds to itself for future
        // withdrawals.

        IECO(_l1Token).transferFrom(_from, address(this), _amount);
        // gons move across the bridge, with inflation multipliers on either side to correctly scale balances
        _amount = _amount * inflationMultiplier;
```

Duplicate of #126
